pageserver/src/http/routes.rs

```diff
@@ -225,28 +249,86 @@ async fn timeline_attach_handler(request: Request<Body>) -> Result<Response<Body
     .await
     .map_err(ApiError::from_err)??;
 
-    let mut remote_index_write = get_state(&request).remote_index.write().await;
-
-    let _enter = span.entered(); // entered guard cannot live across awaits (non Send)
-    let index_entry = remote_index_write
-        .timeline_entry_mut(&ZTenantTimelineId {
-            tenant_id,
-            timeline_id,
-        })
-        .ok_or_else(|| ApiError::NotFound("Unknown remote timeline".to_string()))?;
+    let sync_id = ZTenantTimelineId {
+        tenant_id,
+        timeline_id,
+    };
+    let state = get_state(&request);
+    let remote_index = &state.remote_index;
+
+    let existing_remote_timeline = { remote_index.read().await.timeline_entry(&sync_id).cloned() };
```

Comment:
Indeed.

Yet I don't think that holding a lock throughout the whole operation is a good thing to do.
Should we add some extra field in the `state` maybe?

Reply:
I see several options:
1. How long can this take? Maybe it is ok just to hold the lock throughout the whole operation?
2. We can sort of "shard" the index. So it contains Arc<Mutex<..>> entries, so most of the time, we'll block just one timeline instead of the whole index
3. Keep the current approach, but insert some placeholder entry in the index, so concurrent requests will see it and fail concurrent attach requests.
4. We even can make a map of attach requests in the state itself without touching the index, but we'll need a mutex here aswell, and as we know plural of mutex is deadlock :)

Reply:
1. How long can this take? Maybe it is ok just to hold the lock throughout the whole operation?

The main issue is the `try_download_shard_data` network operation and the related FS manipulations, that could take arbitrary time or fail. The rest of the code could indeed be ok with the index held.

I've rewritten the code so, that we keep the `.write()` lock for longer and release it only to download the `IndexPart` if the corresponding timeline entry is missing.
Looks ok now, since even if another timeline was added, we can just recheck its download status under the `write()` lock and do our scheduling.